### PR TITLE
[#75] add locking mechanism to timestamp generation.

### LINKED ISSUE #75: Timestamp generation is not thread safe
**Example project structure (attached in zip)**
* buildnumbertest
-- A
-- B
-- C
-- D
-- E

Here, submodules A,B,C,D,E will try to generate the timestamp concurrently, which will result in different timestamps assigned.
In this scenario the root project does not attempt to generate the timestamp.

Run the follwing command on the root:
  mvn install -T4 -X -pl A,B,C,D,E | tee out.txt
  
then: 

  cat out.txt | grep "Storing timestamp"

Outputs (which shows that different projects may have different plugins assigned):

[DEBUG] Storing timestamp property: api.test.datetime 09.150811.939
[DEBUG] Storing timestamp property: api.test.datetime 09.150811.955

Note: this is not a deterministic output. Occasionally there is only one timestamp generated.

**Investigation:**

In the org.codehaus.mojo.build.CreateTimestampMojo class, the block of code, is not synchronized:

        if ( timestampString != null )
        {
            getLog().debug( "Using previously created timestamp." );
            return;
        }

        timestampString = Utils.createTimestamp( timestampFormat, timezone );

        getLog().debug( "Storing timestamp property: " + timestampPropertyName + " " + timestampString );

        for ( MavenProject project : session.getProjectDependencyGraph().getSort
[buildnumbertest.zip](https://github.com/mojohaus/buildnumber-maven-plugin/files/1988279/buildnumbertest.zip)
edProjects() )
        {
            getLog().debug( "Storing timestamp property in project " + project.getId() );
            project.getProperties().setProperty( timestampPropertyName, timestampString );
        }

### PR DESCRIPTION
**Pull request by bmarwell**
fixes #75

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -83,7 +83,7 @@
   <properties>
     <maven.scm.version>1.9.5</maven.scm.version>
     <maven.api.version>3.0</maven.api.version>
-    <mojo.java.target>1.7</mojo.java.target>
+    <mojo.java.target>1.8</mojo.java.target>
```
Comment (bmarwell) [inline pom.xml:86]: I also bumped to Java 8.

File: src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java [hunk 2]
```diff
@@ -107,12 +114,38 @@ public void execute()
...
+      if ( lock != null && !currentThreadName.equals(lock))
+      {
+        return false;
+      }
+
+      // lock is either null or set to this thread's name.
+      session.getTopLevelProject().getProperties().setProperty(TIMESTAMP_PROPERTY_LOCK, currentThreadName);
+
+      return true;
+    }
```
Comment (bmarwell) [inline src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:145]: I'm mis-using properties to save a locking property (and remove it afterwards).

File: src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package org.codehaus.mojo.build;
+packag org.codehaus.mojo.build;
```
Comment (olamy) [inline src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:1]: ??

File: src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java [hunk 4]
```diff
@@ -106,12 +112,38 @@ public void execute()
 
         timestampString = Utils.createTimestamp( timestampFormat, timezone );
 
-        getLog().debug( "Storing timestamp property: " + timestampPropertyName + " " + timestampString );
-
-        for ( MavenProject project : session.getProjectDependencyGraph().getSortedProjects() )
+        if (getLock())
```
Comment (olamy) [inline src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:115]: if not getting lock we do not write the timestamp anymore?

**Review by olamy — CHANGES_REQUESTED** [review]

File: src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java (hunk 3 above)
Comment (bmarwell) [reply src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:1]: Oops 😅 how did that happen? Will fix it right away

File: src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java (hunk 4 above)
Comment (bmarwell) [reply src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:115]: How can this happen?

Comment (olamy) [reply src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:116]: looking at the code `getLock() ` can return `false` :)

Comment (bmarwell) [reply src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:116]: Yes, but I think one thread will always get a lock, which is sufficient. That's why I was asking.
Can you picture a scenario where NO thread could get a lock?

Comment (bmarwell) [reply src/main/java/org/codehaus/mojo/build/CreateTimestampMojo.java:116]: @olamy ping

**Pull request closed without merging**